### PR TITLE
Added Reset button to world_control

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

## Summary

Added a reset button to the world Control plugin
reset_button

## Test

```bash
gz sim rolling_shapes.sdf
```

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by mjcarroll** [comment]
A super pedantic nit:  Can we get a reset icon? :)

File: src/plugins/world_control/WorldControl.cc [hunk 1]
```diff
@@ -306,6 +306,18 @@ void WorldControl::OnPause()
   this->dataPtr->SendEventMsg(msg);
 }
 
+/////////////////////////////////////////////////
+void WorldControl::OnReset()
+{
+  gz::msgs::WorldControl msg;
+  auto msgReset = new gz::msgs::WorldReset();
```
Comment (jennuine) [inline src/plugins/world_control/WorldControl.cc:313]: nit: already in the `gz` namespace
```suggestion
  msgs::WorldControl msg;
  auto msgReset = new msgs::WorldReset();
```

File: src/plugins/world_control/WorldControl.qml [hunk 2]
```diff
@@ -82,6 +92,24 @@ RowLayout {
    */
   property bool paused: false
 
+  /**
+   * Reset
+   */
+  RoundButton {
+    id: stopButton
+    objectName: "stopButton"
+    visible: showReset
```
Comment (jennuine) [inline src/plugins/world_control/WorldControl.qml:101]: Should reset only be visible if the sim is paused?

File: src/plugins/world_control/WorldControl.qml [hunk 3]
```diff
@@ -82,6 +92,24 @@ RowLayout {
    */
   property bool paused: false
 
+  /**
+   * Reset
+   */
+  RoundButton {
```
Comment (jennuine) [inline src/plugins/world_control/WorldControl.qml:98]: Can we add a hovered tooltip that states what reset does?

**Review by jennuine — COMMENTED** [review]
> Can we get a reset icon?
+1, the square is misleading. At a glace, it looks like a stop button. 

Also, since we are feature frozen, this will need to come in a minor release

**Comment by ahcorde** [comment]
> A super pedantic nit: Can we get a reset icon? :)

@mjcarroll I was thinking about adding an icon but I was trying to follow the same style which is an unicode geometric shape https://en.wikipedia.org/wiki/List_of_Unicode_characters. I'm open to suggestion.

Anyhow the way that I implemented the `reset` is more like a `stop` (reset + pause)

**Comment by jennuine** [comment]
How about using the same icon as refresh in `ImageDisplay`? https://github.com/gazebosim/gz-gui/blob/a60cb7d61339b407ff3285ddbf16f618c4c4b3b9/src/plugins/image_display/ImageDisplay.qml#L63

**Comment by ahcorde** [comment]
> How about using the same icon as refresh in `ImageDisplay`?
> 
> https://github.com/gazebosim/gz-gui/blob/a60cb7d61339b407ff3285ddbf16f618c4c4b3b9/src/plugins/image_display/ImageDisplay.qml#L63

nice! thank you @jennuine

**Comment by jennuine** [comment]
Users may accidentally click this icon so another nice feature would be to have a confirmation dialog before resetting similar to https://github.com/gazebosim/gz-gui/pull/225 (with the ability to disable the dialog through the config)

**Review by chapulina — COMMENTED** [review]
It would be nice to also add the `Ctrl+R` shortcut, and then we can close

* https://github.com/gazebosim/gz-sim/issues/203

**Comment by scpeters** [comment]
@osrf-jenkins run tests please

**Comment by ahcorde** [comment]
> It would be nice to also add the `Ctrl+R` shortcut, and then we can close
> 
> * [gazebosim/gz-sim#203](https://github.com/gazebosim/gz-sim/issues/203)

https://github.com/gazebosim/gz-gui/pull/476/commits/dbd8c74f31839dd4186bc074e2d0c30ed4748f1c

File: src/plugins/world_control/WorldControl.cc (hunk 1 above)
Comment (ahcorde) [reply src/plugins/world_control/WorldControl.cc]: https://github.com/gazebosim/gz-gui/pull/476/commits/dbd8c74f31839dd4186bc074e2d0c30ed4748f1c

File: src/plugins/world_control/WorldControl.qml (hunk 3 above)
Comment (ahcorde) [reply src/plugins/world_control/WorldControl.qml:98]: https://github.com/gazebosim/gz-gui/pull/476/commits/639643dc602fa71686806d2188548c14f254461b

File: src/plugins/world_control/WorldControl.qml (hunk 2 above)
Comment (ahcorde) [reply src/plugins/world_control/WorldControl.qml:101]: you might want to reset when the simulation is running

**Comment by ahcorde** [comment]
friendly ping @jennuine @mjcarroll

**Review by jennuine — COMMENTED** [review]
nit: can we leave the play/step buttons how they were and add the reset button to the right of the step button? And have the reset button be the same size as step?

**Comment by ahcorde** [comment]
@jennuine 

what do you think about this ?

Selection_094

**Review by mjcarroll — APPROVED** [review]

**Comment by jennuine** [comment]
> @jennuine
> 
> what do you think about this ?
> 
> Selection_094

Looks great, thanks! How do you and @mjcarroll feel about the reset button being to the right of the step button? Reasons being:
1. Users are less likely to accidentally hit reset when attempting to pause
2. It keeps the old/previous layout the same while adding the new feature

**Comment by ahcorde** [comment]
> > @jennuine
> > what do you think about this ?
> > Selection_094
> 
> Looks great, thanks! How do you and @mjcarroll feel about the reset button being to the right of the step button? Reasons being:
> 
> 1. Users are less likely to accidentally hit reset when attempting to pause (I know this is less likely but I can still see it happening)
> 2. It keeps the old/previous layout the same while adding the new feature

Anyhow when you click in the `reset` button a Dialog will appers to ensure that you want to reset the simulation. Not dangerous at all.

**Comment by ahcorde** [comment]
> > > @jennuine
> > > what do you think about this ?
> > > Selection_094
> > 
> > 
> > Looks great, thanks! How do you and @mjcarroll feel about the reset button being to the right of the step button? Reasons being:
> > 
> > 1. Users are less likely to accidentally hit reset when attempting to pause (I know this is less likely but I can still see it happening)
> > 2. It keeps the old/previous layout the same while adding the new feature
> 
> Anyhow when you click in the `reset` button a Dialog will appers to ensure that you want to reset the simulation. Not dangerous at all.

changing layout doesn't seem to be working, the step botton uses a mouseclick area, it's tricky to change it

**Comment by mjcarroll** [comment]
> the step botton uses a mouseclick area, it's tricky to change it

If it is too difficult to fix, I think that the dialog confirmation is sufficiently safe.

**Comment by ahcorde** [comment]
@osrf-jenkins run tests please!

**Comment by ahcorde** [comment]
@osrf-jenkins run tests please!

**Comment by ahcorde** [comment]
@osrf-jenkins run tests please!

**Pull request merged**